### PR TITLE
feat(spacetemplates): Provide pre-configured templates for creating spaces

### PR DESCRIPTION
**Pull request by abelhoula**
## What this PR does / why we need it:
This PR introduces a new feature for SpaceTemplates in the Nauticus project. It enables Spaces to reference SpaceTemplates, allowing users to create Spaces based on predefined templates for easier configuration. The SpaceTemplate feature enhances flexibility and standardization in Space creation.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## How Has This Been Tested?
I've thoroughly tested this feature by creating Spaces that reference SpaceTemplates with various configurations. I've verified that the Spaces inherit the correct settings from the templates and can override them when necessary. Additionally, I've ensured that all unit and integration tests pass successfully.

## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/edixos/nauticus/blob/main/docs/contributing/process.md) guide.
- [x] All new and existing tests passed.

## Does my pull request need a release note?
This PR introduces a new feature for SpaceTemplates, allowing Spaces to reference SpaceTemplates. A release note is required.

```release-note
This PR adds support for SpaceTemplates, enabling Spaces to reference predefined templates for easier configuration. Users can now create Spaces based on these templates to streamline and standardize the Space creation process.

### CODE REVIEW

File: api/v1alpha1/space_types.go [hunk 1]
```diff
@@ -29,6 +29,20 @@ type SpaceSpec struct {
 	LimitRanges LimitRangesSpec `json:"limitRanges,omitempty"`
 	// Specifies a list of service account to create within the Space. Optional
 	ServiceAccounts ServiceAccountsSpec `json:"serviceAccounts,omitempty"`
+	// Reference to a SpaceTemplate
+	Template SpaceTemplateReference `json:"template,omitempty"`
```
Comment (smileisak) [inline api/v1alpha1/space_types.go:33]: it should be `TemplateRef` and refer to json `templateRef`.

File: api/v1alpha1/space_types.go [hunk 2]
```diff
@@ -29,6 +29,20 @@ type SpaceSpec struct {
...
+}
+
+// SpaceTemplateReference
+type SpaceTemplateReference struct {
+	// Name of the SpaceTemplate.
+	Name string `json:"name,omitempty"`
+	// Kind specifies the kind of the referenced resource, which should be "SpaceTemplate".
+	Kind string `json:"kind,omitempty"`
+	// Namespace of the SpaceTemplate.
+	Namespace string `json:"namespace,omitempty"`
```
Comment (smileisak) [inline api/v1alpha1/space_types.go:43]: the `SpaceTemplate` should be a clusterscoped resource as well.

**Review by smileisak — CHANGES_REQUESTED** [review]

File: config/crd/patches/webhook_in_spacetemplates.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# The following patch enables a conversion webhook for the CRD
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: spacetemplates.nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/crd/patches/webhook_in_spacetemplates.yaml:5]: This should be updated. We have made a mistake when we generated the kubebuilder project.

File: config/rbac/role.yaml [hunk 4]
```diff
@@ -91,6 +91,32 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/role.yaml:95]: Same as above

File: PROJECT [hunk 5]
```diff
@@ -13,4 +17,13 @@ resources:
   kind: Space
   path: github.com/edixos/nauticus/api/v1alpha1
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
+  domain: nauticus.io
+  group: nauticus.io
```
Comment (smileisak) [inline PROJECT:25]: The group should be empty I think for next generation.

File: config/rbac/spacetemplate_editor_role.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    app.kubernetes.io/name: clusterrole
+    app.kubernetes.io/instance: spacetemplate-editor-role
+    app.kubernetes.io/component: rbac
+    app.kubernetes.io/created-by: nauticus
+    app.kubernetes.io/part-of: nauticus
+    app.kubernetes.io/managed-by: kustomize
+  name: spacetemplate-editor-role
+rules:
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/spacetemplate_editor_role.yaml:15]: Same as above

File: config/rbac/spacetemplate_editor_role.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  verbs:
+  - create
+  - delete
+  - get
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/spacetemplate_editor_role.yaml:27]: Same

File: config/rbac/spacetemplate_viewer_role.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    app.kubernetes.io/name: clusterrole
+    app.kubernetes.io/instance: spacetemplate-viewer-role
+    app.kubernetes.io/component: rbac
+    app.kubernetes.io/created-by: nauticus
+    app.kubernetes.io/part-of: nauticus
+    app.kubernetes.io/managed-by: kustomize
+  name: spacetemplate-viewer-role
+rules:
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/spacetemplate_viewer_role.yaml:15]: Same

File: config/rbac/spacetemplate_viewer_role.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  resources:
+  - spacetemplates
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/spacetemplate_viewer_role.yaml:23]: Same

**Review by smileisak — CHANGES_REQUESTED** [review]

File: PROJECT (hunk 5 above)
Comment (abelhoula) [reply PROJECT:25]: for now, we keep it like it's?

File: config/crd/patches/cainjection_in_spacetemplates.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
+  name: spacetemplates.nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/crd/patches/cainjection_in_spacetemplates.yaml:7]: Update this to `spacetemplates.nauticus.io`.

File: config/rbac/role.yaml [hunk 11]
```diff
@@ -91,6 +91,32 @@ rules:
...
+  - update
+  - watch
+- apiGroups:
+  - nauticus.io
+  resources:
+  - spacetemplates/finalizers
+  verbs:
+  - update
+- apiGroups:
+  - nauticus.io.nauticus.io
```
Comment (smileisak) [inline config/rbac/role.yaml:113]: Change to `nauticus.io`.

File: controllers/shared/reconsiler.go
Comment (smileisak) [inline controllers/shared/reconsiler.go:1]: reconciler.go instead of reconsiler.go

**Review by smileisak — CHANGES_REQUESTED** [review]
Need to work on the documentation as well.

**Pull request merged**